_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+For example, you can put almost anything in parentheses or branches of the ternary operator, as a part
+of a bigger expression, even if its type cannot be named or even assigned to anything.
+
+#### `using for`
+Since operators are a purely syntactic element, you can always use ordinary functions as a substitute.
+Solidity has a mechanism that is often used this way:
+[`using for`](https://docs.soliditylang.org/en/latest/contracts.html#using-for).
+
+The directive `using ... for ...` can be used to attach functions to a type.
+These functions can then be used as members on any object of that type and receive the object as
+their first parameter.
+
+```solidity
+pragma solidity ^0.8.13;
+
+type Int is int;
+using {add} for Int global;
+
+function add(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function test(Int a, Int b) pure returns (Int) {
+    return a.add(b);
+}
+```
+
+Currently, `using for` has 3 forms:
+1. `using L for T`: attaches all non-private functions from library `L` to type `T`.
+1. `using L for *`: attaches all non-private functions from library `L` to any type, including types that
+    cannot be named in the language and used explicitly (e.g. types of literals or array slices).
+1. `using {f, L.g, ...} for T`: attaches given free or library functions `f`, `L.g`, ... to type `T`.
+
+The directive can be used:
+- ✅ At the file level.
+- ✅ Inside contract and library definitions.
+
+Can not be used:
+- ❌ Inside interface definitions.
+
+When used inside contracts, it is not inherited.
+Its effects extend only to the base contract that has it and not to any other contracts derived from it.
+
+At the file level, the directive affects all the free functions, contracts and libraries defined within that file.
+It does not affect other files that import the file containing it, unless it is marked as `global`.
+`global` extends its effect to all files where the type is visible.
+The use of `global` is possible only with UDVTs, structs, and enums that are defined in its source unit.
+
+### Introducing User-Defined Operators
+The new feature extends the third form of the `using for` syntax to allow binding operators to functions.
+The function is invoked whenever the operator is used on the values of the type:
+
+```solidity
+pragma solidity ^0.8.19;
+
+type Int is int;
+using {add as +} for Int global;
+
+function add(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function test(Int a, Int b) pure returns (Int) {
+    return a + b; // Equivalent to add(a, b)
+}
+```
+
+An operator can only be bound in a `global` directive and the definition must be a `pure` free function.
+The type specified in `using for` must be a user-defined value type and must be the type
+of all the parameters of the function and its return value.
+The only exception are comparison operators, where the return type must be `bool`.
```

Comment:
Similar to the docs in the actual PR, I'd prefer to say that each operator has a unique signature and to list them... but if you think that table is too much for here, we can keep it like this.

Reply:
I think it's fine in the context of a blog post. It's an accurate summary of the current situation and the post is written from the perspective of this point in time, no one will expect it to be updated with the language.

And yeah, I think including the whole table would make it too long (the post is already very long). I really want this bit to be concise.

Reply:
My problem is that to me the phrasing suggests a conceptually wrong view on the operators :-). There is only one global comparison operator. It's always the same. It always returns ``bool``. You just implement it for types.
The current phrasing suggests that you have multiple unrelated operators flying around and then impose restrictions on their types.
Of course, in the end observationally the result is the same, but it's quite different conceptualizations.

Reply:
But yeah, as I said originally, we can keep it, if need be :-).

Reply:
> The current phrasing suggests that you have multiple unrelated operators flying around and then impose restrictions on their types.
Of course, in the end observationally the result is the same, but it's quite different conceptualizations.

Fair enough. What if I say `comparison operator definitions`? Because that's what I really meant. I probably should have been more precise about this. Currently you *can* have "multiple unrelated operator *definitions* flying around".